Prevent `UnitOfWork` lookup for DBAL types specified in `Doctrine\ORM\Query#setParameter()`

Ocramius opened the pull request:
Fixes #7527

Ocramius committed:
#7527 failing test case: `UnitOfWork#getSingleIdentifierValue()` should not be called for a well specified parameter type

As previously reported by @flaushi in https://github.com/doctrine/doctrine2/pull/7471#discussion_r241949045, we discovered
that binding a parameter causes a `ClassMetadataFactory#getClassMetadata()` call, which in turn leads to large performance
regression when using any `object` type as parameter.

Following two snippets lead to an internal `ClassMetadataFactory#getClassMetadata()` call, which in turn leads to an
exception being thrown and garbage collected, plus multiple associated performance implications:

```php
$query->setParameter('foo', new DateTime());
$query->getResult();
```

```php
$query->setParameter('foo', new DateTime(), DateTimeType::NAME);
$query->getResult();
```

This is due to following portion of code:

https://github.com/doctrine/doctrine2/blob/434820973cadf2da2d66e7184be370084cc32ca8/lib/Doctrine/ORM/Query.php#L406-L409

Notice how `$value = $this->processParameterValue($value);` happens before attempting to infer the type for the parameter value.

That call leads to this segment being reached, which leads to the regression:

https://github.com/doctrine/doctrine2/blob/434820973cadf2da2d66e7184be370084cc32ca8/lib/Doctrine/ORM/AbstractQuery.php#L423-L433

Assuming the bound parameter type is provided, we can completely skip attempting to introspect the given object:

```php
$query->setParameter('foo', new DateTime(), DateTimeType::NAME);
$query->getResult();
```

Processing the parameter value is not needed in this case, so we can safely skip that logic for all known parameters.
In order to not introduce a BC break or change the `AbstractQuery#processParameterValue()` implementation, we could filter
out all parameters for which the type is given upfront, and later on merge them back in instead.

The test expectation to be set is for `UnitOfWork#getSingleIdentifierValue()` to never be called.

Ocramius committed:
Note: this will still lead to the `UnitOfWork#getSingleIdentifierValue()` still being
called when not specifying the type of a DQL parameter being bound via
`Doctrine\ORM\Query#setParameter()`:

```php
$query->setParameter('foo', $theValue, $theType);
```

A full parameter bind is required in order to gain back performance:

```php
$query->setParameter('foo', $theValue, $theType);
```

This is up for discussion with patch reviewers.

Ocramius commented on lib/Doctrine/ORM/Query/Parameter.php:
```diff
@@ -49,6 +49,13 @@ class Parameter
+     *
+     * @var bool
+     */
+    private $typeSpecified;
```
I'm not 100% happy about this, but so far, the constructor of `Parameter` is destructive (this information is lost), hence this solution allows us to preserve the initial scope.

Changing the constructor is not viable, as it will lead to even more BC breaks.

greg0ire commented on lib/Doctrine/ORM/Query.php:
```diff
@@ -439,6 +428,38 @@ private function processParameterMappings($paramMappings)
+        $value         = $originalValue;
+        $rsm           = $this->getResultSetMapping();
+
+        assert($rsm !== null);
```
I love assertions over `@var`, but Is it really useful in this case? The phpdoc doesn't say anything about `null` being a possibility:

https://github.com/doctrine/doctrine2/blob/2f7508b3e337d695e44815666d2a7ad287874e54/lib/Doctrine/ORM/AbstractQuery.php#L456

Is this about auto-completion? Phpstan? Both?

Ocramius replied:
Not sure, just added it here due to what was being inferred.

lcobucci reviewed (dismissed):
@Ocramius code looks good, what do you think about also adding a bench to check set a baseline for queries with and without defined types?

Ocramius commented:
I'm wondering if there is an actual way to prevent the regression without defined types though...

lcobucci commented:
> I'm wondering if there is an actual way to prevent the regression without defined types though...

@Ocramius the only way I'd imagine is to create a fitness function based on the benchmark results and assert that they are equal to a baseline (with some minor acceptable delta). However this would require a reliable and standardised environment to run them (at least manually), which we don't have at the moment.

Ocramius commented:
Nono, that's something I can unit test (besides the performance bench, which I'll do later): I don't have a technical solution to prevent processing parameters with inferred types.

lcobucci commented:
> Nono, that's something I can unit test (besides the performance bench, which I'll do later): I don't have a technical solution to prevent processing parameters with inferred types.

@Ocramius my point is mostly related to the fact that the bug fixed bug @alcaeus makes us always try to load the metadata of an object, which is the (only?) way we currently have to check if we're dealing with an entity or not. The optimisation point then (IMHO) is the validation of whether the given value is a potential entity or not, which is 100% valid but unrelated to this PR and might open a can of worms - one that is kind of being tackled in `v3.0`.

What am I not seeing?

Ocramius commented:
I'm just wondering if I can recover from the performance regression in this scenario (no specified type):

```php
$query->setParameter('foo', new DateTime());
```

Not sure if this is possible, and patch doesn't touch this, but I'm open for suggestions if anybody sees a possible architectural solution.

lcobucci commented:
@Ocramius we had the previous performance because `$this->_em->getMetadataFactory()->hasMetadataFor(ClassUtils::getClass($value))` was doing an early return in https://github.com/doctrine/doctrine2/pull/7471#discussion-diff-241949045R415. But that is exactly the fix done in #7471 since some objects didn't have their metadata loaded yet and were being used directly  as parameter.

As far as I can see, the performance issues are now happening because we're going till the metadata driver and trying to load the metadata of an object that is NOT an entity only to ignore the exception and return the original object.

The only think I can think of is to verify if an object is a potential entity or not without relying on class metadata factory - removing the need for throwing an unused exception (which also has performance implications).

Ocramius commented:
> verify if an object is a potential entity or not without relying on class metadata

Which is a bit of an oxymoron: the only thing telling us if an object is an entity or not *is* class metadata.

I think we can finalise the patch as-is and close it here. I'll try adding the benchmarks later today.

lcobucci commented:
> Which is a bit of an oxymoron: the only thing telling us if an object is an entity or not is class metadata.

I know, we can always add a nasty white list fated to rot through the years :rofl:

Ocramius committed:
#7527 performance benchmark - verifying performance impact of inferred query parameter types

As an example result:

```
./phpbench.phar run tests/Doctrine/Performance/Query --iterations=50 --revs=50 --report=aggregate
PhpBench 0.15-dev (dcbe193). Running benchmarks.
Using configuration file: /home/ocramius/Documents/doctrine/doctrine2/phpbench.json

\Doctrine\Performance\Query\QueryBoundParameterProcessingBench

    benchExecuteParsedQueryWithInferredParameterTypeI49 P0 	[μ Mo]/r: 643.684 634.664 (μs) 	[μSD μRSD]/r: 17.700μs 2.75%
    benchExecuteParsedQueryWithDeclaredParameterTypeI49 P0 	[μ Mo]/r: 97.673 94.251 (μs) 	[μSD μRSD]/r: 8.259μs 8.46%

2 subjects, 100 iterations, 100 revs, 0 rejects, 0 failures, 0 warnings
(best [mean mode] worst) = 88.460 [370.679 364.458] 127.400 (μs)
⅀T: 37,067.880μs μSD/r 12.980μs μRSD/r: 5.603%
suite: 133f0e30090f815142331ebec6af18241694e7c0, date: 2018-12-19, stime: 10:47:10
+------------------------------------+--------------------------------------------------+--------+--------+------+-----+------------+-----------+-----------+-----------+-----------+----------+--------+-------+
| benchmark                          | subject                                          | groups | params | revs | its | mem_peak   | best      | mean      | mode      | worst     | stdev    | rstdev | diff  |
+------------------------------------+--------------------------------------------------+--------+--------+------+-----+------------+-----------+-----------+-----------+-----------+----------+--------+-------+
| QueryBoundParameterProcessingBench | benchExecuteParsedQueryWithInferredParameterType |        | []     | 50   | 50  | 5,970,568b | 604.680μs | 643.684μs | 634.664μs | 677.640μs | 17.700μs | 2.75%  | 6.59x |
| QueryBoundParameterProcessingBench | benchExecuteParsedQueryWithDeclaredParameterType |        | []     | 50   | 50  | 5,922,424b | 88.460μs  | 97.673μs  | 94.251μs  | 127.400μs | 8.259μs  | 8.46%  | 1.00x |
+------------------------------------+--------------------------------------------------+--------+--------+------+-----+------------+-----------+-----------+-----------+-----------+----------+--------+-------+
```

This indicates that the performance impact for NOT declaring parameter types
explicitly is *MASSIVE*.

Ocramius commented:
@lcobucci this is ready to :ship: 

The merge is intentional, since I was moving back and forth with the benchmark branch.

lcobucci commented:
@Ocramius Travis says no, though :stuck_out_tongue: I'll review it in a bit.

Ocramius commented:
Yeah, CS issues, can fix myself later

alcaeus reviewed (commented):
Changes look good from my end, I'll let @lcobucci give his seal of approval 👍

Ocramius committed:
#7527 automated CS checks

lcobucci reviewed (approved):
Let's get this merged in, it's quite good. Thanks @Ocramius

X-Coder264 commented:
@lcobucci @Ocramius When is 2.6.4 gonna be tagged with this fix? Unfortunately we are "stuck" on 2.6.2 until this gets tagged as it breaks our application in the queries where we are binding DateTime objects (as the code on 2.6.3 now  reaches our custom MappingDriver implementation and it throws an exception there). Seeing how this has such an impact is there any reason why 2.6.4 couldn't be tagged as soon as possible?
